_posts/2020-04-27-tworzenie-i-usuwanie-indeksow-w-bazie-mongodb-na-produkcji.md

```diff
@@ -0,0 +1,79 @@
+
+Procedurę zaczynamy od węzłów secondary, a na końcu przeprowadzamy ją dla węzła primary.
+
+## Odłączenie węzła od klastra
+
+Odłączenie węzła od klastra wymaga 2 kroków:
+- zmiany w konfiguracji, aby węzeł po restarcie nie podłączył się do klastra,
+- restartu węzła.
+
+W konfiguracji musimy zmienić port, wykomentować konfigurację repliki oraz wyłączyć odświeżanie sesji. Przykładowo fragment pliku z konfiguracją może wyglądać tak:
+```yaml
+net:
+  bindIp: 127.0.0.1
+  port: 28017
+#   port: 27017
+#replication:
+#   replSetName: rs0
+
+# Wyłącz odświeżanie sesji
+setParameter:
+   disableLogicalSessionCacheRefresh: true
+```
+
+Uwaga! Jeżeli węzeł jest primary, to musimy wykonać operację _step down_ przed restartem, aby stał się secondary, a klaster wybrał inny węzeł jako primary. Służy temu polecenie `rs.stepDown()`.
+
+Po zmianach w konfiguracji restartujemy węzeł. W tym momencie węzeł jest odłączony od klastra i działa w trybie standalone. Teraz możemy utworzyć indeks.
+
+## Tworzenie indeksu
+
+Po restarcie węzła i uruchomieniu w trybie standalone, podłączamy się do niego i wykonujemy polecenie tworzące indeks zgodnie z [dokumentacją](https://docs.mongodb.com/manual/reference/method/db.collection.createIndex/). Tworzenie indeksu może okazać się czasochłonne. Aktualnie wykonywane operacje na bazie sprawdzimy, uruchamiając polecenie `db.currentOp()` w powłoce MongoDB. To pozwoli dowiedzieć się, czy indeks nadal się buduje.
+
+## Dołączenie węzła do klastra
+
+Po zbudowaniu indeksu dołączamy węzeł do klastra. Polega to na wycofaniu zmian w konfiguracji oraz restarcie węzła. W tym momencie węzeł będzie się aktualizował względem klastra (węzła primary), czyli pobierał zmiany, które zaszły w czasie kiedy był odłączony.
+
+Informacje o klastrze są zwracane przez polecenia:
+- `rs.status()` - zwraca status klastra względem węzła, na którym polecenie zostało uruchomione;
+- `rs.printSlaveReplicationInfo()` - zwraca informacje o węzłach secondary łącznie z opóźnieniem względem węzła primary.
+
+# Usuwanie indeksu
+
+Usuwanie indeksu jest łatwiejszą operacją. Wystarczy usunąć indeks na węźle primary. Całą resztą, czyli usunięciem indeksu z węzłów secondary, zajmie się MongoDB. Musimy jednak się upewnić, że żadne zapytanie nie korzysta z danego indeksu. Inaczej zapytanie zakończy się błędem.
```

Comment:
Tutaj nie jest tak różowo. To co piszesz jest prawdą od 4.2. Wcześniej (w szczególnosci w wersji 4.0, o której piszesz na początku) dropIndex ubije wszystkie zapytania bez wyjątku na tej kolekcji. Być może, jeśli zależ nam na bezprzerwowej pracy systemu, to dropa też należy robić w trybie rolling?

Reply:
Poprawiłem